### PR TITLE
feat: Update the NodeJS Dockerfilegenerator transformer and Dockerfile template to support multi-arch build

### PR DESCRIPTION
**Pull request by Akash-Nayak**
Signed-off-by: Akash Nayak <akash19nayak@gmail.com>

### CODE REVIEW

**Comment by Akash-Nayak** [comment]
Hi @ashokponkumar ji, I have updated this PR to use `ubi8/nodejs` image in the Dockerfile template for NodeJS and also build the image and tested the deployment on OpenShift cluster.

New Dockerfile generated by Move2Kube.
```console
FROM registry.access.redhat.com/ubi8/nodejs-14
COPY . .
RUN npm install
RUN npm run build
USER root
RUN chown -R 1001:0 /opt/app-root/src/.npm
RUN chmod -R 775 /opt/app-root/src/.npm
USER 1001
EXPOSE 8080
CMD npm run start
```


Currently, I have not removed the image-tag logic that I added, and I have commented it out in case if we want to switch to docker hub [Node](https://hub.docker.com/_/node/tags) image in future.

File: transformer/dockerfilegenerator/nodejsdockerfiletransformer.go [hunk 1]
```diff
@@ -99,15 +101,18 @@ type NodejsDockerfileGenerator struct {
 
 // NodejsDockerfileYamlConfig represents the configuration of the Nodejs dockerfile
 type NodejsDockerfileYamlConfig struct {
-	DefaultNodejsVersion  string `yaml:"defaultNodejsVersion"`
-	DefaultPackageManager string `yaml:"defaultPackageManager"`
+	DefaultNodejsVersion string `yaml:"defaultNodejsVersion"`
+	// DefaultNodejsTag          string            `yaml:"defaultNodejsTag"`
+	DefaultPackageManager   string            `yaml:"defaultPackageManager"`
+	NodejsVersionTagMapping map[string]string `yaml:"nodejsVersionTagMapping"`
```
Comment (ashokponkumar) [inline transformer/dockerfilegenerator/nodejsdockerfiletransformer.go:107]: Aren't we taking this as a config file?

File: transformer/dockerfilegenerator/nodejsdockerfiletransformer.go [hunk 2]
```diff
@@ -99,15 +101,18 @@ type NodejsDockerfileGenerator struct {
...
 }
 
 const (
 	packageJSONFile        = "package.json"
 	versionMappingFilePath = "mappings/nodeversions.yaml"
 	defaultPackageManager  = "npm"
-	versionKey             = "version"
+	// tagKey                 = "tag"
```
Comment (ashokponkumar) [inline transformer/dockerfilegenerator/nodejsdockerfiletransformer.go:114]: what is tag?

File: assets/built-in/transformers/dockerfilegenerator/nodejs/templates/Dockerfile [hunk 3]
```diff
@@ -12,17 +12,7 @@
...
-    {{- else }}
-    wget https://nodejs.org/dist/{{ .NodeVersion }}/node-{{ .NodeVersion }}-linux-x64.tar.xz && \
-    {{- end }}
-    tar -xJf node-{{ .NodeVersion }}-linux-x64.tar.xz && \
-    mv node-{{ .NodeVersion }}-linux-x64 /node && \
-    rm -f node-{{ .NodeVersion }}-linux-x64.tar.xz
-ENV PATH="$PATH:/node/bin"
+FROM registry.access.redhat.com/ubi8/nodejs-{{ .NodeMajorVersion }}
 COPY . .
 RUN {{ .PackageManager }} install
```
Comment (ashokponkumar) [inline assets/built-in/transformers/dockerfilegenerator/nodejs/templates/Dockerfile:17]: Does the image have both npm and yarn?

File: transformer/dockerfilegenerator/nodejsdockerfiletransformer.go [hunk 4]
```diff
@@ -249,9 +267,11 @@ func (t *NodejsDockerfileGenerator) Transform(newArtifacts []transformertypes.Ar
...
 		}
 		nodejsConfig := NodejsTemplateConfig{
-			Build:                 build,
-			Port:                  port,
-			NodeVersion:           nodeVersion,
+			Build:       build,
+			Port:        port,
+			NodeVersion: nodeVersion,
+			// NodeTag:               nodeTag,
+			NodeMajorVersion:      strings.TrimPrefix(semver.Major(nodeVersion), "v"),
```
Comment (ashokponkumar) [inline transformer/dockerfilegenerator/nodejsdockerfiletransformer.go:274]: We might not have all the node versions in the images, isn't it? So we might want to have a list of versions that are present and go to the nearest that fits the version criteria.

**Review by ashokponkumar — CHANGES_REQUESTED** [review]
Nice work. Few comments.

File: transformer/dockerfilegenerator/nodejsdockerfiletransformer.go [hunk 5]
```diff
@@ -135,8 +140,16 @@ func (t *NodejsDockerfileGenerator) Init(tc transformertypes.Transformer, env *e
 	t.Spec = spec
 	if t.NodejsConfig.DefaultNodejsVersion == "" {
 		t.NodejsConfig.DefaultNodejsVersion = t.Spec.NodeVersions[0][versionKey]
```
[the remark this answers is not in the document]
Comment (Akash-Nayak) [reply transformer/dockerfilegenerator/nodejsdockerfiletransformer.go]: Done.

File: transformer/dockerfilegenerator/nodejsdockerfiletransformer.go (hunk 1 above)
Comment (Akash-Nayak) [reply transformer/dockerfilegenerator/nodejsdockerfiletransformer.go]: I have removed this.

File: assets/built-in/transformers/dockerfilegenerator/nodejs/templates/Dockerfile (hunk 3 above)
Comment (Akash-Nayak) [reply assets/built-in/transformers/dockerfilegenerator/nodejs/templates/Dockerfile:20]: The ubi8/nodejs-{version} base image only has npm and doesn't have yarn. Earlier in the ubi8-minimal image, we were downloading node but not installing yarn. So, I have added the command to install yarn if the PackageManager is yarn. https://github.com/Akash-Nayak/move2kube/blob/03df5e77bd6734eb18dfd6645b3c108549655203/assets/built-in/transformers/dockerfilegenerator/nodejs/templates/Dockerfile#L18

File: transformer/dockerfilegenerator/nodejsdockerfiletransformer.go (hunk 4 above)
Comment (Akash-Nayak) [reply transformer/dockerfilegenerator/nodejsdockerfiletransformer.go:263]: Yes, currently there are 4 node versions in the ubi8/node images (10, 12, 14 and 16) that can be pulled. 

```console
docker run --rm -it -p 8080:8080 registry.access.redhat.com/ubi8/nodejs-10 bash
bash-4.4$ node -v
v10.24.0
```

```console
docker run --rm -it -p 8080:8080 registry.access.redhat.com/ubi8/nodejs-12 bash
bash-4.4$ node -v
v12.22.5
```

```console
docker run --rm -it -p 8080:8080 registry.access.redhat.com/ubi8/nodejs-14 bash
bash-4.4$ node -v
v14.20.1
```

```console
docker run --rm -it -p 8080:8080 registry.access.redhat.com/ubi8/nodejs-16 bash
bash-4.4$ node -v
v16.17.1
```

We have the updated list of available versions here- https://github.com/Akash-Nayak/move2kube/blob/13141078de0de81e2c705d85b31f162d06772933/assets/built-in/transformers/dockerfilegenerator/mappings/nodeversions.yaml#L6 and we select the node version that fits the version criteria here https://github.com/Akash-Nayak/move2kube/blob/03df5e77bd6734eb18dfd6645b3c108549655203/transformer/dockerfilegenerator/nodejsdockerfiletransformer.go#L222.

**Review by ashokponkumar — APPROVED** [review]

**Pull request merged**